File: datasets/lua-keep-incoder_6b_temp_0.2-1/HumanEval_10_make_palindrome.lua
-- Find the shortest palindrome that begins with a supplied string.
-- Algorithm idea is simple:
-- - Find the longest postfix of supplied string that is a palindrome.
-- - Append to the end of the string reverse of a string prefix that comes before the palindromic suffix.
-- >>> make_palindrome('')
-- ''
-- >>> make_palindrome('cat')
-- 'catac'
-- >>> make_palindrome('cata')
-- 'catac'
local function make_palindrome(string)
  local palindrome = string
  local postfix = string
  local postfix_len = #string
  local postfix_reverse = string
  local postfix_reverse_len = #string
  local palindrome_len = #palindrome
  while postfix_len > 0 do
    local postfix_reverse = string
    local postfix_reverse_len = #string
    for i = 1, postfix_len do
      local char = string.sub(postfix, i, i)
      local char_reverse = string.sub(postfix_reverse, i, i)
      if char_reverse ~= char then
        postfix_reverse = string.sub(postfix, 1, i-1) .. char_reverse .. string.sub(postfix, i+1, i)
        postfix_reverse_len = #postfix_reverse
      end
    end
    if postfix_reverse ~= postfix then
      postfix = postfix_reverse
      postfix_len = postfix_reverse_len
    end
  end
  return palindrome
end

-- Find the longest palindrome that begins with a supplied string.
-- Algorithm idea is simple:
-- - Find the longest postfix of supplied string that is a palindrome.
-- - Append to the end of the string reverse of a string prefix that comes before the palindromic suffix.
-- >>> make_palindrome('')
-- ''
-- >>> make_palindrome('cat')
-- 'catac'
-- >>> make_palindrome('cata')
-- 'catac'

lu = require('luaunit')

function test_humaneval()
local candidate = make_palindrome
    lu.assertEquals(candidate(''), '')
    lu.assertEquals(candidate('x'), 'x')
    lu.assertEquals(candidate('xyz'), 'xyzyx')
    lu.assertEquals(candidate('xyx'), 'xyx')
    lu.assertEquals(candidate('jerry'), 'jerryrrej')
end

os.exit(lu.LuaUnit.run())

File: datasets/R-keep-code_davinci_001_temp_0.2/HumanEval_10_make_palindrome.R
# Find the shortest palindrome that begins with a supplied string.
-- Algorithm idea is simple:
-- - Find the longest postfix of supplied string that is a palindrome.
-- - Append to the end of the string reverse of a string prefix that comes before the palindromic suffix.
-- >>> make_palindrome('')
-- ''
-- >>> make_palindrome('cat')
-- 'catac'
-- >>> make_palindrome('cata')
-- 'catac'
make_palindrome <- function(string)
{
    if (is.null(string))
    {
        return(NULL)
    }
    
    if (length(string) == 0)
    {
        return('')
    }
    
    if (length(string) == 1)
    {
        return(string)
    }
    
    if (length(string) == 2)
    {
        if (string[1] == string[2])
        {
            return(string)
        }
        else
        {
            return(paste(string, string[1], sep = ''))
        }
    }
    
    # Find the longest postfix of supplied string that is a palindrome.
    palindrome <- NULL
    for (i in 1:length(string))
    {
        if (is_palindrome(string[i:length(string)]))
        {
            palindrome <- string[i:length(string)]
        }
    }
    
    # Append to the end of the string reverse of a string prefix that comes before the palindromic suffix.
    return(paste(string, reverse(palindrome), sep = ''))

test_humaneval <- function() {
candidate <- make_palindrome
    if(!identical(candidate(''), '')){quit('no', 1)}
    if(!identical(candidate('x'), 'x')){quit('no', 1)}
    if(!identical(candidate('xyz'), 'xyzyx')){quit('no', 1)}
    if(!identical(candidate('xyx'), 'xyx')){quit('no', 1)}
    if(!identical(candidate('jerry'), 'jerryrrej')){quit('no', 1)}
}